Input Profile Overhaul

iwubcode opened the pull request:
Input Profiles are vital to playing more than one or two wii games with emulated controllers.

The current profile system works but has a number of short comings:

- Having more than a handful of profiles and you quickly will need to organize them.  The current combo-box only lists profiles by name and assumes they will all be in that specific directory.  Sub-folders are not supported which is useful when needing to organize by game and then by device.
- The current OSD message that occurs when a profile is loaded is just the profile filename without extension, it gives you no context if a profile is loaded that has the same name as another profile (that exists in a different sub-directory)
- Trying to cycle profiles in multiple locations, won't work because the profile selected isn't stored on the controller (I ran into this when working on #7930 )

This PR addresses these issues.

MayImilae commented:
Hmm. So the profile never had any sort of "active profile" indicator because profiles are never active. Like, loading a profile just takes what is in the profile and uses it to insta-fill the configuration UI, so the profile is only used for an instant.

If there is an "current profile" indicator, I think we'll run into issues of users thinking that adjustments should be auto-saved to the profile, as though the profile is constantly being read/written. Personally I think that downside outweighs the convenience of that indicator.

lioncash commented on Source/Core/InputCommon/ControllerEmu/ControllerEmu.cpp:
```diff
@@ -96,6 +105,21 @@ void EmulatedController::LoadConfig(IniFile::Section* sec, const std::string& ba
+
+    std::string profile_path;
+    sec->Get(base + "ProfilePath", &profile_path, "");
+    if (profile_path != "")
```
```suggestion
    if (!profile_path.empty())
```

lioncash commented on Source/Core/InputCommon/ControllerEmu/ControllerEmu.h:
```diff
@@ -4,6 +4,8 @@
 
 #pragma once
 
+#include "ProfileManager.h"
```
This should be a relative path and be down with the other InputCommon include.

lioncash commented on Source/Core/InputCommon/ControllerEmu/ProfileManager.cpp:
```diff
@@ -0,0 +1,124 @@
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "ProfileManager.h"
```
This include should be relative to the project root, i.e. `InputCommon/ControllerEmu/ProfileManager.h`

lioncash commented on Source/Core/InputCommon/ControllerEmu/ProfileManager.h:
```diff
@@ -0,0 +1,44 @@
+
+#pragma once
+
+#include "InputCommon/DeviceProfileContainer.h"
```
These should be below the standard includes, since it's a header

lioncash commented on Source/Core/InputCommon/ControllerEmu/ProfileManager.h:
```diff
@@ -0,0 +1,44 @@
+class ProfileManager
+{
+public:
+  explicit ProfileManager(ControllerEmu::EmulatedController* owning_controller);
```
You'd likely want to default the destructor in the cpp file as well, given the class contains a non-trivial type (std::set)

lioncash commented on Source/Core/InputCommon/DeviceProfileContainer.cpp:
```diff
@@ -0,0 +1,110 @@
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "DeviceProfileContainer.h"
```
This should be a relative include

lioncash commented on Source/Core/InputCommon/DeviceProfileContainer.h:
```diff
@@ -0,0 +1,32 @@
+
+#pragma once
+
+#include "InputProfile.h"
```
This should be a relative include and be below the other includes.

lioncash commented on Source/Core/InputCommon/InputProfile.h:
```diff
@@ -1,53 +1,30 @@
-  void PreviousWiimoteProfile(int controller_index);
-  void NextWiimoteProfileForGame(int controller_index);
-  void PreviousWiimoteProfileForGame(int controller_index);
+  explicit InputProfile(const std::string& profile_path);
```
This should probably have the destructor defaulted in the cpp file, given the class contains entirely non-trivial types.

lioncash commented on Source/Core/InputCommon/InputProfile.h:
```diff
@@ -1,53 +1,30 @@
-class InputConfig;
+#include "Common/IniFile.h"
+
+#include <string>
```
This should be above the IniFile include.

lioncash commented on Source/Core/InputCommon/InputProfile.h:
```diff
@@ -1,53 +1,30 @@
-// Copyright 2018 Dolphin Emulator Project
+// Copyright 2019 Dolphin Emulator Project
```
This doesn't need to be changed when modifications are made to the file. It's intended to establish a creation date.

lioncash commented on Source/Core/InputCommon/InputProfile.cpp:
```diff
@@ -1,210 +1,32 @@
-// Copyright 2018 Dolphin Emulator Project
+// Copyright 2019 Dolphin Emulator Project
```
Ditto

lioncash commented on Source/Core/InputCommon/DeviceProfileContainer.cpp:
```diff
@@ -0,0 +1,110 @@
+{
+namespace
+{
+int clamp_container(int index, std::size_t container_size)
```
```suggestion
int ClampContainer(int index, std::size_t container_size)
```

iwubcode commented:
Thank you @lioncash for reviewing!

@MayImilae - That makes sense.  I will remove the profile indicator after I finish my testing.  Thank you for the feedback.

iwubcode commented:
This PR has no addressed all feedback.  I've done testing with a variety of scenarios and have not seen any issues.  Please feel free to test add, delete, save of profiles in addition to various profile loading from a game-setting or hotkey scheduler.

BhaaLseN commented on Source/Core/InputCommon/ControllerEmu/ControllerEmu.cpp:
```diff
@@ -111,6 +120,21 @@ void EmulatedController::LoadConfig(IniFile::Section* sec, const std::string& ba
+      {
+        return;
+      }
+      // TODO: warn
```
You didn't forget about this `TODO`, did you?

iwubcode commented:
Unlikely I will be getting back to this any time soon, closing.
